datafusion/src/physical_plan/expressions/variance.rs

```diff
@@ -0,0 +1,376 @@
+            | DataType::UInt64
+            | DataType::Int8
+            | DataType::Int16
+            | DataType::Int32
+            | DataType::Int64
+            | DataType::Float32
+            | DataType::Float64
+    )
+}
+
+impl Variance {
+    /// Create a new VARIANCE aggregate function
+    pub fn new(
+        expr: Arc<dyn PhysicalExpr>,
+        name: impl Into<String>,
+        data_type: DataType,
+    ) -> Self {
+        // the result of variance just support FLOAT64 data type.
+        assert!(matches!(data_type, DataType::Float64));
+        Self {
+            name: name.into(),
+            expr,
+        }
+    }
+}
+
+impl AggregateExpr for Variance {
+    /// Return a reference to Any that can be used for downcasting
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn field(&self) -> Result<Field> {
+        Ok(Field::new(&self.name, DataType::Float64, true))
+    }
+
+    fn create_accumulator(&self) -> Result<Box<dyn Accumulator>> {
+        Ok(Box::new(VarianceAccumulator::try_new()?))
+    }
+
+    fn state_fields(&self) -> Result<Vec<Field>> {
+        Ok(vec![
+            Field::new(
+                &format_state_name(&self.name, "count"),
+                DataType::UInt64,
+                true,
+            ),
+            Field::new(
+                &format_state_name(&self.name, "mean"),
+                DataType::Float64,
+                true,
+            ),
+            Field::new(
+                &format_state_name(&self.name, "m2"),
+                DataType::Float64,
+                true,
+            ),
+        ])
+    }
+
+    fn expressions(&self) -> Vec<Arc<dyn PhysicalExpr>> {
+        vec![self.expr.clone()]
+    }
+
+    fn name(&self) -> &str {
+        &self.name
+    }
+}
+
+/// An accumulator to compute variance
+#[derive(Debug)]
+pub struct VarianceAccumulator {
+    m2: ScalarValue,
+    mean: ScalarValue,
+    count: u64,
+}
+
+impl VarianceAccumulator {
+    /// Creates a new `VarianceAccumulator`
+    pub fn try_new() -> Result<Self> {
+        Ok(Self {
+            m2: ScalarValue::from(0 as f64),
+            mean: ScalarValue::from(0 as f64),
+            count: 0,
+        })
+    }
+
+    pub fn get_count(&self) -> u64 {
+        self.count
+    }
+
+    pub fn get_mean(&self) -> ScalarValue {
+        self.mean.clone()
+    }
+
+    pub fn get_m2(&self) -> ScalarValue {
+        self.m2.clone()
+    }
+}
+
+impl Accumulator for VarianceAccumulator {
+    fn state(&self) -> Result<Vec<ScalarValue>> {
+        Ok(vec![
+            ScalarValue::from(self.count),
+            self.mean.clone(),
+            self.m2.clone(),
+        ])
+    }
+
+    fn update(&mut self, values: &[ScalarValue]) -> Result<()> {
+        let values = &values[0];
+        let is_empty = values.is_null();
+
+        if !is_empty {
+            let new_count = self.count + 1;
+            let delta1 = ScalarValue::add(values, &self.mean.arithmetic_negate())?;
```

Comment:
Using `ScalarValue`s like this to accumulate each operation is likely to be quite slow during runtime. However, I think it would be fine to put in as a first initial implementation and then implement an optimized version using `update_batch` and arrow compute kernels as a follow on PR

```rust
    /// updates the accumulator's state from a vector of arrays.
    fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
```

Reply:
Completely agree. I have plan to investigate this as followup PR. The current challenge is the algorithm will loose parallelizability if using a batch friendly algorithm. And I need to spend more time to understand the code. One question I have is will there be a chance `update` and `batch_update` can be used in the same job, i.e. if one job can call `update` on some data and `batch_update` on some other data. Reason for that is the online version of the algorithm requires an intermediate value to be calculated so it is not compatible with batch mode, that is, we can only do all batch or all online.

Reply:
I don't fully understand the discussion here about  online and parallelism with respect to `update` and `update_batch`

`update_batch` is actually what the HashAggregator calls  during execution. The default implementation of `update_batch` simply iterates over each row of the input array, converting it to a `ScalarValue`, and calls `update` convert the  https://github.com/apache/arrow-datafusion/blob/415c5e124af18a05500514f78604366d860dcf5a/datafusion/src/physical_plan/mod.rs#L572-L583

My mental picture of aggregation is like this:

```
┌───────────────────────┐             ┌──────────────────────┐                                       
│                       │             │                      │                                       
│  Input RecordBatch 0  │ update_batch│     Accumulator      │                                       
│                       │────────────▶│                      │───────┐                               
│                       │             │                      │       │                               
└───────────────────────┘             └──────────────────────┘       │                               
                                                                     │                               
                                                                     │     merge                     
                                                                     │                               
┌───────────────────────┐             ┌──────────────────────┐       │       ┌──────────────────────┐
│                       │             │                      │       │       │                      │
│  Input RecordBatch 1  │ update_batch│     Accumulator      │       │       │     Accumulator      │
│                       │────────────▶│                      │───────┼──────▶│                      │
│                       │             │                      │       │       │                      │
└───────────────────────┘             └──────────────────────┘       │       └──────────────────────┘
                                                                     │                               
                                                                     │                               
                                                                     │                               
                                                                     │                               
           ...                                   ...                 │                               
                                                                     │                               
                                                                     │                               
                                                                     │                               
┌───────────────────────┐             ┌──────────────────────┐       │                               
│                       │             │                      │       │                               
│  Input RecordBatch N  │ update_batch│     Accumulator      │       │                               
│                       │────────────▶│                      │───────┘                               
│                       │             │                      │                                       
└───────────────────────┘             └──────────────────────┘                                       
```

Reply:
Sorry let me try to explain it a bit better.

The online algorithm uses an intermediate value for each record called m2. Here is how m2 is calculated:
`m2(n) = m2(n-1) + (delta1*delta2)`
So the calculation of m2(n) depends on the value of m2(n-1). And it is similar for delta1 and delta2 - they depends on mean(n-1). i.e. the algorithm is iterative. So what I meant was this algorithm can't benefit from complete vectorization - for each batch it has to be iterative. 

But I see your point that we can avoid per-record type conversion and matching since in `batch_update` we can use the raw array. I think I can make the change in this PR. Thanks for pointing it out.

Reply:
>  So what I meant was this algorithm can't benefit from complete vectorization - for each batch it has to be iterative.

Ah that makes sense 👍